source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,863 @@
+to error reporting.
+
+Deprecations
+~~~~~~~~~~~~
+
+The following configuration timeout options MUST be deprecated in favor of
+``timeoutMS``:
+
+- ``socketTimeoutMS``
+
+- ``waitQueueTimeoutMS``
+
+- ``wTimeoutMS``
+
+The following options for CRUD methods MUST be deprecated in favor of
+``timeoutMS``:
+
+- ``maxTimeMS``
+
+- ``maxCommitTimeMS``
+
+Drivers MUST throw an error during MongoClient configuration if ``timeoutMS``
+is used in conjunction with one of the options in the deprecated list. This
+validation MUST NOT be done during MongoDatabase or MongoCollection
+configuration.
+
+Timeout Behavior
+----------------
+
+The ``timeoutMS`` option specifies the best-effort maximum amount of time a
+single operation can take before control is returned to the application.
+Drivers MUST keep track of the remaining time before the timeout expires as
+an operation progresses.
+
+Operations
+~~~~~~~~~~
+
+The ``timeoutMS`` option applies to all operations defined in the following
+specifications:
+
+- `CRUD <./../crud/crud.rst>`__
+- `Change Streams <../change-streams/change-streams.rst>`__
+- `Client Side Encryption <../client-side-encryption/client-side-encryption.rst>`__
+- `Enumerating Collections <../enumerate-collections.rst>`__
+- `Enumerating Databases <../enumerate-databases.rst>`__
+- `GridFS <../gridfs/gridfs-spec.rst>`__
+- `Index Management <../index-management.rst>`__
+- `Transactions <../transactions/transactions.rst>`__
+
+In addition, it applies to all operations on cursor objects that may perform
+blocking work (e.g. methods to iterate or close a cursor, any method that
+reads documents from a cursor into an array, etc).
+
+Validation and Overrides
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+When executing an operation that supports the ``timeoutMS`` option, drivers
+MUST raise a validation error if ``timeoutMS`` is set on the operation or is
+inherited from the collection/database/client levels in conjunction with one
+of the deprecated timeout options. In addition to being set at these levels,
+the timeout for an operation can also be expressed via an explicit
+ClientSession (see `Convenient Transactions API`_). In this case, the timeout
+on the session MUST be used as the ``timeoutMS`` value for the operation.
+Drivers MUST raise a validation error if an explicit session with a timeout
+is used and the ``timeoutMS`` option is set at the operation level.
+
+Errors
+~~~~~~
+
+If the ``timeoutMS`` option is not set and support for deprecated timeout
+options has not been dropped but a timeout is encountered (e.g. server
+selection times out), drivers MUST continue to return existing errors. This
+ensures that error-handling code in existing applications does not break
+unless the user opts into using ``timeoutMS``.
+
+If the ``timeoutMS`` option is set and the timeout expires, drivers MUST
+abort all blocking work and return control to the user with an error. This
+error MUST be distinguished in some way (e.g. custom exception type) to make
```

Comment:
> Is it still required to distinguish it?

Yes. Ideally, a user should be able to check for a single exception type (e.g. `TimeoutException`).

> Can the new error be a subtype of the existing one?

Yes, I think that's fine. I wasn't prescriptive about this in the spec, though, because the concept of wrapping errors is different across languages. In Java, it should be sufficient to declare a new `MongoTimeoutException` type and have the existing `MongoServerSelectionTimeoutException` be a subtype of it.

Reply:
In java driver we already throw a somewhat generic MongoTimeoutException for server selection and connection pool timeouts.  Is this saying we need a different type, or can we continue to use the same type?

Reply:
If there's already a generic type, it seems fine to continue using that as long as other timeout errors (e.g. socket timeouts) are changed to be a subtype of it.